flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/TimeGauge.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// ----------------------------------------------------------------------------
+//  This class is largely adapted from "com.google.common.base.Preconditions",
+//  which is part of the "Guava" library.
+//
+//  Because of frequent issues with dependency conflicts, this class was
+//  added to the Flink code base to reduce dependency on Guava.
+// ----------------------------------------------------------------------------
+
+package org.apache.flink.runtime.metrics.groups;
+
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.metrics.Gauge;
+import org.apache.flink.metrics.View;
+
+/**
+ * {@link TimeGauge} encapsulate logic of measuring time rates (like idleTime ms/s). In particularly
+ * it takes care of the case, when single measurement (as defined by {@link #markStart()} and {@link
+ * #markEnd()}), crosses the boundary of single {@link #update()} interval. Thanks to the {@link
+ * #markStart()} {@link TimeGauge} can correctly calculate next value in {@link #update()}, even if
+ * this measurement hasn't not ended ({@link #markEnd()}).
```

Comment:
It is not obvious from this description why this is important.

Overall I would wager that it would be difficult to accurately state what this thing does based on the javadocs.
Maybe something alone the lines of "...measures how much time is spent in a given state, with entry into that state being signaled by #markStart()" could help with that.
Then it also makes it obvious why it is important to account for the case that markEnd has not been called yet.

Reply:
Looks good 👍